Enable non-static schema registration

jeffbloo opened the pull request:
This changes how operator schema are registered and used by ONNX. 

The motivation for this change is two-fold.  First, it transfers policy to the consumer for which operator schema and versions of those schema will be registered.  Second, it avoids a pattern of static object usage which can cause downstream compilation and performance issues.

Before this change, schema are registered using a macro (ONNX_OPERATOR_SCHEMA) which populates a static map.  This PR changes these macros to implement methods constructing the schema.  It also exposes functions that iterate across schema in different operator sets.

By default, the static registration map is populated with the operator sets on first use.  Callers can also disable this and populate the map manually.  Finally they can implement ISchemaRegistry, and request this be  which is used instead of the static map.

So that consumers may generate code with calls to the new registration functions, the macros are now always called from the ONNNX namespace.  The codebase was previously inconsistent on using either this or the global namespace.  The objects defined in this macro also use a naming convention, which includes the name, domain, and version which reflects the operator set within which it was first defined.  This affects the macro syntax and removes the need to use __COUNTER__ for uniqueness.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/onnx/onnx?pullRequest=894) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/onnx/onnx?pullRequest=894) before we can accept your contribution.<br/><hr/>**Jeff Bloomfield** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/onnx/onnx?pullRequest=894) it.</sub>

bddppq commented:
Please resolve merge conflict (and make sure the CI passing).

bddppq commented:
>Second, it avoids a pattern of static object usage which can cause downstream compilation and performance issues.

Could you elaborate?

linkerzhang commented on onnx/defs/schema.h:
```diff
@@ -655,14 +662,21 @@ class OpSchemaRegistry final {
-#define ONNX_OPERATOR_SCHEMA_UNIQ(Counter, name)                 \
+// ONNX_OPERATOR_SCHEMA is implemented either with static registration or by exposing
+// a factory method.  
+#ifndef ONNX_NON_STATIC_SCHEMA_INITIALIZATION
```
I'm nervous about adding this kind of option, similar with the one we added for ONNX_ML. it makes the code diverge and different partners focus on each option. It's generally hard to maintain. For such divergence, either all partners agree on same direction, or put the specific stuff in local code base.

shschaefer replied:
I agree that having different patterns is not ideal.  However, not all vendors may be able to leverage the code in its current state, but the code contains a lot of necessary information.  To answer the question above @bddppq, control of the timing of initialization is important to manage potential startup cost for the compilation unit.  As we are looking at this, the code can be brought into many different processes, some where deferring this cost is important.

shschaefer replied:
For non-framework scenarios, when this code is loaded as part of an ONNX library, it can be consumed by numerous processes.  We do a lot of work to eliminate this overhead, delay initialization until the library is used, etc...  The dynamic initializer makes this code hard to consume.  Not sure if providing benchmarks changes this.

bddppq replied:
@shschaefer I'm asking benchmark numbers mainly because I think the overhead of these static initialization should be negligible.

jeffbloo replied:
I measured that registering schema consumes 4 milliseconds and increases memory consumption by 1.1 MB.  This is very non-negligible for a single component, especially if it may be loaded into many processes.

bddppq commented on onnx/defs/schema.h:
```diff
@@ -655,14 +662,21 @@ class OpSchemaRegistry final {
+template <typename T> OpSchema GetOpSchema();
+#define ONNX_OPERATOR_SCHEMA(name, domain, ver, impl) \
+    class name##_##domain##_ver##ver##_OpSchemaId{}; \
+    template<> OpSchema ONNX_NAMESPACE::GetOpSchema< name##_##domain##_ver##ver##_OpSchemaId>(){ \
```
Actually could you elaborate how does it work? Where is this `GetOpSchema` declared and how is user code supposed to call it?

jeffbloo replied:
GetOpSchema is declared slightly above this.  Macro call sites define its specializations.  

Based on the feedback here, I'm preparing a change to remove the original mode of registration during static initialization.  This will include a new file with related declarations for each schema, and functions which retrieve and/or register the schema of specific operator sets.

bddppq commented:
@jeffbloo @linkerzhang @shschaefer @dzhulgakov How about we move the schema instantiation from global static variables to a function static variable in the `map()` function? i.e. something like:
```
class SchemasRegisterer {
  SchemasRegisterer() {
   // AUTOGENERATED BEGIN
    _map.insert(std::make_pair("Aaa", OperatorSchema(Aaa)....));
    ...
   _map.insert(std::make_pair("Zzz", OperatorSchema(Zzz)....));
   // AUTOGENERATED END
  }
}

Map* map() {
  static auto SchemasRegisterer{};
  return _map
}
```
This way we only do the initialization at the first usage time, and there is no need for an Init and it's thread safe.

To achieve this we will need to move the source of truth of the schemas to say yaml files (instead .cc files) and then have a script to auto-generate the chunk between `AUTOGENERATED BEGIN` and `AUTOGENERATED END` above.

jeffbloo commented:
@bddppq - It's possible to populate the map by using the same couple code lines as added to cpp2py_export.cc.  Is it your goal to make this change non-breaking, so that updates are not required in any callers?

Another downside to that change would be coupling the policy of whether to use OpSchemaRegistry with the policy of which operator sets and versions will be used, in which registry.  

This could also make it less clear whether other code in the ONNX library should rely on this map.  For a similar reason, I was considering removing OpSchemaRegistry's internal statics altogether, and making callers create their own static instances of OpSchemaRegistry and share them as needed.  This is more disruptive to calling code though.

I'd prefer to keep this simple and define operator sets directly for now.  Code generation could be a more involved change if designed holistically (considering the most efficient format to generate, how to best handle inferencing callbacks, etc.).   The iteration routines in the existing PR also provide more flexibility for the library consumer, without requiring them to use a code-gen step.

jeffbloo commented:
The latest update works around a bug in older GCC compilers that the build hit.  Raw strings inside macros were not handled correctly.

I also removed usages of std::abort to avoid affecting terminating a whole process, and added a conditional to allow large doc strings to be compiled out to save memory.

bddppq commented on onnx/defs/experiments/defs.cc:
```diff
@@ -212,6 +207,11 @@ ONNX_OPERATOR_SCHEMA(GivenTensorFill, ONNX_DOMAIN, 1, OpSchema()
         }
     }));
 
+const char * Scale_ver1_doc = R"DOC(
```
What's the reason of moving this string outside? This makes the schema (in c++ files) less readable.

jeffbloo replied:
This is the workaround I mentioned for a GCC bug.  Raw strings in macros were busted.
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=57824

bddppq commented on onnx/defs/schema.h:
```diff
@@ -659,14 +704,27 @@ class OpSchemaRegistry final {
+
+// Defines specialization of GetOpSchema for a class whose name is determined based on 
+// a convention using name, domain, and version.
+#define ONNX_OPERATOR_SCHEMA(name, domain, ver, impl) \
```
most schemas are of the onnx core domain. can you change this to a helper macro and leave `ONNX_OPERATOR_SCHEMA` dedicated to the onnx core domain? For ONNX_ML I think `ONNX_ML_OPERATOR_SCHEMA` can be used.

jeffbloo replied:
I can do that

bddppq commented on onnx/defs/schema.cc:
```diff
@@ -312,62 +331,59 @@ OpSchema& OpSchema::Attr(
-    a.set_type(attr_type);                                          \
-    Attr(Attribute(std::move(name), std::move(description), a));    \
-    return *this;                                                   \
+#define ATTR_SETTER_WITH_SINGLE_COMPLEXVALUE(type, field, attrtype)       \
```
Could you not clang-format the unchanged part? It's really hard to review. :-)

jeffbloo replied:
OK

bddppq commented on onnx/defs/schema.cc:
```diff
@@ -5,6 +5,8 @@
 #include <unordered_set>
 #include "onnx/checker.h"
+#include "onnx/defs/operator_sets.h"
+#include "onnx/defs/operator_sets-ml.h"
```
Need to guard this include with ONNX_ML for non ONNX_ML builds

jeffbloo replied:
The headers had internal conditionals, but I can add them here.

bddppq commented on onnx/defs/operator_sets.h:
```diff
@@ -0,0 +1,385 @@
+// Copyright (c) Facebook Inc. and Microsoft Corporation.
```
So this file should be autogenerated, it's so easy to forget to update when adding new schemas and also unfriendly for new contributors.

jeffbloo replied:
In either case, the operators won't be available or get generated into documentation, so I don't think this will be missed.  

Not requiring code-gen steps is simplifying in itself.  While I can imagine other benefits if we combine code-gen with other refactoring, I don't want to block this PR on that larger project.

bddppq replied:
This will be missed. We can't rely on reviewers to remember how many docs changes they need to see so the changes are legit, the only viable way to guard such boilerplate changes is adding a check in the CI. Also think about new coming/occasional contributors, it's impossible for them to remember.

jeffbloo replied:
Our thinking was that exposing operators which were never tested was not as important, but I understand wanting to make this easier to discover.

I'll see if I can cleanly add check in debug mode that the number of registration macros usages matches the total of schema exposed by the iteration methods.  Does that sound reasonable?

jeffbloo replied:
I've made that change.  This also fixes a pre-existing bug in assert implementations which caused them to always pass.

bddppq commented on onnx/defs/schema.cc:
```diff
@@ -575,11 +591,32 @@ std::ostream& operator<<(std::ostream& out, const OpSchema& schema) {
+      }
+  };
+
+  #ifndef ONNX_DISABLE_STATIC_REGISTRATION
```
Can we remove this variant? It's not happening at initialization time already, what's the concern?

jeffbloo replied:
This was intended to allow the static registry to be used by a caller which wants to constrain it to specific operator sets.  For example, to deprecate an older operator set or withhold others.

jeffbloo replied:
This applies to your question below about custom schema. They may also implement ISchemaRegistry.

bddppq replied:
Could you elaborate what's the purpose of constrain the registration to a specific operator sets? If it's for overriding an operator definition in the onnx core I'm pretty against it.

jeffbloo replied:
No problem.  It's not actually intended for overriding ones that already exist.  One goal is for a caller to have more control over the contract it exposes to its own customers.  This naturally includes the operators that it supports. 

Another goal is to enable callers to handle certain operator sets outside the core library, based in part on its strategy for deploying updates.

bddppq commented on onnx/defs/schema.h:
```diff
@@ -209,10 +210,25 @@ class OpSchema final {
-  OpSchema& SetDoc(std::string doc);
+  // This may be disabled to save memory.
+  OpSchema& SetDoc(const char* doc) {
+#ifndef ONNX_NO_DOC_STRINGS
```
Can you separate the doc string stripping change to anther PR? I prefer not to let these two discussions don't slow down each other. Also according to our previous discussion in #623, removing doc strings really doesn't save you too much (~50kb **uncompressed** size, <10kb **compressed**)

jeffbloo replied:
I see a similar gain - 56 KB.  I also have to make the strings static and avoid consuming onnx.lib with whole-archive (which is now possible with this change).  I think this is worthwhile for such an isolated change.  Savings like this one add up.

jeffbloo replied:
This applies to your question below about custom schema.  They may also implement ISchemaRegistry.

bddppq replied:
What's the total size in your measurement? If it's the same as in the previous PR, the total size is around 8MB, which means 50 KB is really not worth adding this complexity into the code base (note as you have found this not only needs to add those macros, but also require moving couple strings and changing some function signature in the OpSchema class, and in some code you will need to add guards preventing accessing the doc strings).
In any case, I feel this really should be a separate discussion, could you split it out into a separate PR.

jeffbloo replied:
The baseline size was around 2.5 MB.  In the context of a larger product this isn't as relevant.  Components of different sizes sharing a common budget should be equally judicious with size deltas, and this is an easy win.   It also reduces heap usage, which is multiplied across every single process loading the binary.

Doc strings are optional, so guards aren't necessary.  The string movements were necessary for different reasons - the GCC bug workaround.  The signature changes also also needed for other reasons - minimizing static STL string initialization and avoiding instruction bloat in the macro-defined functions.

bddppq commented:
Also what's the strategy for users adding their own schemas? As mentioned before, moving the schemas instantiation into function static variable makes the registration closed.

jeffbloo commented:
@bddppq - As to binary size, this was indeed a slight increase.  Measured on our compiler, about 85K larger while optimizing for speed, and 55K for size.  This is without using the new macro to remove doc strings.

A lot of bloat is related to constructing STL objects before each registration call.  I'm fixing this by simply wrapping the OpSchema methods consuming STL types.  This will make the size 175K smaller than the master branch, and another 56K smaller if removing doc strings.

bddppq commented on onnx/checker.cc:
```diff
@@ -465,7 +464,7 @@ void check_model(const ModelProto& model) {
   } else {
     if (opset_imports.empty())
-      opset_imports[ONNX_DOMAIN] = 1;
+      opset_imports[ONNX_DOMAIN_NAME] = 1;
```
Could you preserve ONNX_DOMAIN as well?

bddppq commented on onnx/common/ir_pb_converter.cc:
```diff
@@ -70,7 +70,7 @@ Tensor tensorProtoToTensor(const ONNX_NAMESPACE::TensorProto & tp) {
   }
   case ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED:
-    abort();
+    throw std::runtime_error("Unknown tensor data type");
```
In other places (e.g. checker and shape inference) we define a custom exception class inherit from std::runtime_error and macros (e.g. `fail_check`, `fail_shape_inference`).
